File: 프로그래머스/3/43163. 단어 변환/단어 변환.py
from collections import deque
def solution(begin, target, words):
    if target not in words:
        return 0
    
    queue = deque([(begin, 0)])
    visited = set()
    
    while queue :
        current_word, steps = queue.popleft()
        if (current_word == target) :
            return steps
        
        for next_word in words:
            if next_word not in visited and checkChangeWord(current_word, next_word) :
                visited.add(next_word)
                queue.append((next_word, steps + 1))
    
    return 0

def checkChangeWord(begin, target) :
    sameChar = 0
    for idx in range(len(begin)) :
        if (begin[idx] != target[idx]) :
            sameChar += 1

    return sameChar == 1
    

File: 프로그래머스/3/43163. 단어 변환/단어 변환.java
import java.util.Queue;
import java.util.LinkedList;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

class Solution {
    
    static class Node {
        String word;
        Integer count;

        public Node(String word, Integer count) {
            this.word = word;
            this.count = count;
        }
        
        public String getWord() {
            return this.word;
        }
        
        public Integer getCount() {
            return this.count;
        }
    }
    
    public int solution(String begin, String target, String[] words) {
        if (!Arrays.asList(words).contains(target)) {
            return 0;
        }
        
        List<String> visited = new ArrayList<>();
        Queue<Node> queue = new LinkedList<>();
        queue.offer(new Node(begin, 0));
        
        while(!queue.isEmpty()) {
            Node pollNode = queue.poll();
            String pollWord = pollNode.getWord();
            Integer pollCount = pollNode.getCount();
            
            if (pollWord.equals(target)) {
                return pollCount;
            }
                
            for (String word : words) {
                if (!Arrays.asList(visited).contains(word) && checkOneWord(pollWord, word)) {
                    visited.add(word);
                    queue.offer(new Node(word, pollCount+1));
                }
            }
        }
        
        return 0;
    }
    
    private Boolean checkOneWord(String word1, String word2) {
        int checkCount = 0;
        for (int i =0; i < word1.length(); i++) {
            if (word1.charAt(i) != word2.charAt(i)) {
                checkCount++;
            }   
        }
        
        return checkCount == 1;
    }
}